### ISSUE TITLE
Strange  failures in JIT compiled code, maybe related to Cache

### ISSUE THREAD

**Issue by th-otto** [aranym/aranym#18]
While trying to hunt down bugs in the JIT compiler, i'm currently getting a bit desperate. There are 2 symptoms currently, one apparently related to Cache management, the other to FPU (or maybe also caused by the same bug).

I first encountered that error when running an example from povray (fish13.pov, which seems to be rather complex, since it takes about 30 min when run on Hatari)

The output from povray when run with JIT+FPU looks like [this](http://tho-otto.de/screenshots/fish13-020.png), but should actually look like [this](http://tho-otto.de/screenshots/fish13-000-fdlibm.png). The output looks ok when i let the compilation of fmove Fn,<ea> fail, so that the normal software emulation is used instead. It is in compemu_fpp.cpp, and all it does is a call to put_fp_value.

The other symptom manifests in more or less random failures from the testsuite i recently created for fdlibm. The strange thing there is, that i usually run a script which just calls all the test executables, and one of the tests (copysign, which is a real simple function) fails. Even more strange, that same test works when i invoke it separately. And as if that would not be enough, it also does not fail when i compile the executable using gcc 8.3. 

So that gave me the idea that something is wrong with the cache management of the JIT compiler. And indeed, when i set the cache size to 0, all the mentioned problems go away. However this cannot be the real solution, the povray example takes about 3min when run without a cache, and about 12sec when run with the default settings (8MB).

Can someone confirm that behaviour? povray executables can be found [here](http://tho-otto.de/crossmint.php), and a stripped down version of the fdlibm test files is attached below.

[fdlibm-small.zip](https://github.com/aranym/aranym/files/3081102/fdlibm-small.zip)

**Comment by mikrosk** [comment]
I guess @andreas-schwab is the local FPU expert. :)

**Comment by andreas-schwab** [comment]
You cannot trust the FPU emulation unless you are using the mpfr emulation, and even then you cannot trust it.  Also, the m68k FPU is not IEEE compliant, so you cannot even trust the real hardware.

**Comment by th-otto** [comment]
Yes, thanks for the hints;) I know that there are some subtle differences especially when using the x87 FPU (different handling of denormals, treating INFs sometimes as NaN if the explicit integer bit is not set, slightly different rounding modes etc). Still those things don't explain the failures, since the a) can be avoided as explained above by just turning the cache off and b) the first failure only occurs with JIT, but not when turning it off and using the IEEE fpu, which in this case essentially does the same operations. And b) the other failure is not related to FPU code at all, it happens with the softfloat version of fdlibm (the failing test there is for copysign, and it looks like copysign(x, y) is returning the value of y instead of x).
For example in line 9 where it does a copysign(0, -4) it returns -4

Of course you can just turn off JIT compilation of FPU instructions completely if you don't trust it, but that was not the purpose of implementing it ;) Also it really makes a huge speed differences especially for programs like povray. Beside that, IEEE emulation is the only choice for the JIT version, since not all instructions are compiled, and the remaining ones are executed using the normal CPU emulation, which then must use a compatible mode.

**Comment by bishopcop** [comment]
For me this problem also manifests in Raystart, even the 3d drawing routines get corrupted, the renderer crashes to the desktop.  The Fpu is not used.  If cache is set to 0 then it works fine. Interestingly if I use MagiC, it seems to work fine also.

**Comment by bishopcop** [comment]
Latest commits seems to have fixed this issue.

**Comment by th-otto** [comment]
Good to know. By the wrong rendering of the fish in povray seems to remain.

**Comment by bishopcop** [comment]
I've tested several versions of POV with the fish scene file.  That weird issue only crops up on the MINT version of POV 3.6, the one you linked.  On POV 3.0 FPU version, it renders correctly, as well as POV 2.2 TT version.

**Comment by mikrosk** [comment]
For what it's worth, I observe *much* better compatibility with the latest snapshots, esp. under Windows/JIT setup but also on Linux. I could steadily reproduce some gcc internal errors when using the JIT version and now those errors are gone.
